### PR TITLE
fix(libsinsp): fix some path handling in fs.path

### PR DESCRIPTION
**Pull request by incertum**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

/kind bug

/kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

> /kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area API-version

> /area build

> /area CI

> /area driver-kmod

> /area driver-bpf

> /area driver-modern-bpf

> /area libscap-engine-bpf

> /area libscap-engine-gvisor

> /area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-savefile

> /area libscap

> /area libpman

/area libsinsp

> /area tests

> /area proposals

**Does this PR require a change in the driver versions?**

> /version driver-API-version-major

> /version driver-API-version-minor

> /version driver-API-version-patch

> /version driver-SCHEMA-version-major

> /version driver-SCHEMA-version-minor

> /version driver-SCHEMA-version-patch

**What this PR does / why we need it**:

`fs.path*` still had some minor bugs in handling path resolution and ensuring no trailing `/` is in the output fields.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
fix(libsinsp): fix some path handling in fs.path
```

### CODE REVIEW

File: userspace/libsinsp/sinsp_filtercheck_fspath.cpp [hunk 1]
```diff
@@ -349,7 +348,14 @@ uint8_t* sinsp_filter_check_fspath::extract(sinsp_evt* evt, OUT uint32_t* len, b
 			return NULL;
 		}
 
-		m_tstr = sinsp_utils::concatenate_paths(tinfo->get_cwd(), m_tstr);
+		if(m_tstr.front() != '/')
+		{
+			m_tstr = sinsp_utils::concatenate_paths(tinfo->get_cwd(), m_tstr.c_str());
```
Comment (incertum) [inline userspace/libsinsp/sinsp_filtercheck_fspath.cpp:353]: @federico-sysdig and @LucaGuerra hope you are ready for the next round :woman_facepalming:
Similar to the last PR we also have some issues here with the string. The possible trailing slash would not be removed otherwise. I have adjusted the unit tests to test this as well. Plus we also need to call `concatenate_paths` when the path is absolute. Absolute paths with path traversal happen on real-life servers.

File: userspace/libsinsp/test/events_fspath.ut.cpp [hunk 2]
```diff
@@ -27,17 +27,21 @@ class fspath : public sinsp_with_test_input
...
+	const char *resolved_rel_filename = "/root/tmp/filename.txt";
 
 	const char *rel_filename_complex = "../\\.../../tmp/filename_complex";
 	const char *resolved_rel_filename_complex = "/tmp/filename_complex";
 
 	const char *rel_filename_nopath = "nopath";
 	const char *resolved_rel_filename_nopath = "/root/nopath";
 
-	const char *name = "/tmp/name";
+	const char *name = "/tmp/random/dir...///../../name/";
```
Comment (incertum) [inline userspace/libsinsp/test/events_fspath.ut.cpp:41]: Test for possible trailing `/` and path traversal for absolute paths.

File: userspace/libsinsp/test/events_fspath.ut.cpp [hunk 3]
```diff
@@ -181,42 +185,10 @@ class fspath : public sinsp_with_test_input
 		sinsp_evt* evt = add_event_advance_ts_v(increasing_ts(), 1, event_type, n, args);
 		va_end(args);
 
-		verify_fields(evt, expected_name, expected_name, NULL, NULL, NULL, NULL);
-	}
-
-	void test_exit_path_raw(const char *expected_name,
```
Comment (incertum) [inline userspace/libsinsp/test/events_fspath.ut.cpp:187]: We could convert `test_exit_path_raw` to `test_exit_path`. That way the unit tests are clearer and more functional.

**Review by incertum — COMMENTED** [review]
To the best of my testing abilities `fd.name` and `fs.path.name` now behave consistently the same way.

File: userspace/libsinsp/sinsp_filtercheck_fspath.cpp [hunk 4]
```diff
@@ -349,7 +348,14 @@ uint8_t* sinsp_filter_check_fspath::extract(sinsp_evt* evt, OUT uint32_t* len, b
...
+		} else
+		{
+			// concatenate_paths takes care of resolving the path
+			m_tstr = sinsp_utils::concatenate_paths("", m_tstr.c_str());
```
Comment (federico-sysdig) [inline userspace/libsinsp/sinsp_filtercheck_fspath.cpp:357]: `std::string` should be automatically cast to `std::string_view`, I don't think you need the `c_str()` call (which needs an automatic cast from `const char*` to `std::string_view` anyway). A similar simplification can be done a few lines above.

**Review by federico-sysdig — COMMENTED** [review]
Wrote just a few comments.

File: userspace/libsinsp/sinsp_filtercheck_fspath.cpp (hunk 4 above)
Comment (incertum) [reply userspace/libsinsp/sinsp_filtercheck_fspath.cpp:357]: We had that discussion before, nothing else worked here ...

File: userspace/libsinsp/utils.cpp [hunk 5]
```diff
@@ -632,7 +632,7 @@ std::string sinsp_utils::concatenate_paths(std::string_view path1, std::string_v
 	//
 	// If the path ends with a separator, remove it, as the OS does.
 	//
-	if (result.length() > 1 && result.back() == '/')
+	if (!result.empty() && result.back() == '/')
```
[the remark this answers is not in the document]
Comment (incertum) [reply userspace/libsinsp/utils.cpp:635]: Can revert later. From my perspective our string handling isn't very transparent (at least to me), when do we expect null terminated etc, in a different place in the other PR we use `.empty()`  in another place. The fact we continue having problems https://github.com/falcosecurity/libs/pull/1571#discussion_r1424954825 means something is off or not consistent throughout our code base. WDYT?

[the remark this answers is not in the document]
Comment (federico-sysdig) [reply userspace/libsinsp/utils.cpp:635]: I don't understand. My comment was not on the technique to check whether a string is null. I was noticing that testing the length for begin at least one character (`str.length() > 1`) isn't exactly the same as testing if the string is not empty (`!str.empty()`). Changing one into the other makes the function behave differently, is this intentional?

File: userspace/libsinsp/sinsp_filtercheck_fspath.cpp (hunk 4 above)
Comment (federico-sysdig) [reply userspace/libsinsp/sinsp_filtercheck_fspath.cpp:357]: Please, forgive me, I forgot. I'm aging. I'm noticing that we do pass a straight `std::string` to this function elsewhere. Do you mean that in this specific call instance it is not working? As in not compiling or not doing the right concatenation?

Comment (incertum) [reply userspace/libsinsp/sinsp_filtercheck_fspath.cpp:357]: It's not doing the right concatenation. I developed and tested on my fedora Linux box.

File: userspace/libsinsp/utils.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (LucaGuerra) [reply userspace/libsinsp/utils.cpp:635]: I agree with @federico-sysdig :)
However, re. @incertum 's uncertainity about string handling, it's an actual problem: strings should behave like strings and not have extra nulls at the end, see the fix in https://github.com/falcosecurity/libs/pull/1574 . If we find other instances of such things (hopefully there aren't) we need to fix them :)

**Comment by incertum** [comment]
Rebased and squashed with the solution we all want, but I am now still getting some errors locally:

```
[  PASSED  ] 417 tests.
[  FAILED  ] 4 tests, listed below:
[  FAILED  ] fspath.unlinkat_2
[  FAILED  ] fspath.open
[  FAILED  ] fspath.openat_2
[  FAILED  ] fspath.openat2
```

all similar like:
```
Expected equality of these values:
  get_field_as_string(evt, fs_path_name).c_str()
    Which is: "/root/tmp/name/"
  expected_name
    Which is: "/root/tmp/name"
```

File: userspace/libsinsp/utils.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (incertum) [reply userspace/libsinsp/utils.cpp]: Apologies yes reverted.

**Comment by incertum** [comment]
CI is confirming these failures.

File: userspace/libsinsp/utils.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (LucaGuerra) [reply userspace/libsinsp/utils.cpp]: > hopefully there aren't

There are :) I will better explain in the comments

**Comment by LucaGuerra** [comment]
@incertum the problem lies in the `m_tstr.assign()` calls in that file. As you can see the are taking a raw buffer (incl. NUL terminator) and converting it into a string by giving it the total length so they will always include the NUL. We don't want that, so we should change those assignments to build the string only up to its length to e.g.:

```c++
m_tstr.assign((const char*) extract_values[0].ptr, strnlen((const char*) extract_values[0].ptr, extract_values[0].len));
```

At this point I think we can safely remove the part that says:

```c++
	// If m_tstr ends in a c-style \0, remove it to be
	// consistent. Generally, the evt.rawarg fields above *will*
	// end in c-style \0 characters, while the ones that extract
	// from the enter event or have values populated by parsers
	// (e.g. fchmod/fchown) will not.
	if(m_tstr.back() == '\0')
	{
		m_tstr.pop_back();
	}
```

This becomes redundant because we already deal with the problem earlier as we should, I think it's good to maintain the invariant that `m_tstr` is a well formed string.

File: userspace/libsinsp/sinsp_filtercheck_fspath.cpp [hunk 6]
```diff
@@ -349,7 +348,14 @@ uint8_t* sinsp_filter_check_fspath::extract(sinsp_evt* evt, OUT uint32_t* len, b
 			return NULL;
 		}
 
-		m_tstr = sinsp_utils::concatenate_paths(tinfo->get_cwd(), m_tstr);
+		if(m_tstr.front() != '/')
```
Comment (LucaGuerra) [inline userspace/libsinsp/sinsp_filtercheck_fspath.cpp:351]: once the problem explained below is fixed I believe we can safely revert this change.

Comment (incertum) [reply userspace/libsinsp/sinsp_filtercheck_fspath.cpp:351]: Follow up: I think we still need this as in this case we concatenate cwd and m_tstr, and else we still use the helper to possibly resolve the path.

**Comment by incertum** [comment]
Re https://github.com/falcosecurity/libs/pull/1571#issuecomment-1854818644 thanks @LucaGuerra :rocket: -- suggestions incorporated plus I learned dos and donts with string handling for the future, thanks for that!

File: userspace/libsinsp/sinsp_filtercheck_fspath.cpp (hunk 6 above)
Comment (FedeDP) [reply userspace/libsinsp/sinsp_filtercheck_fspath.cpp:351]: Can we use `std::filesystem::path::is_absolute` instead? Much more readable IMHO.
If i got this correctly, we are stating that if `m_tstr` is relative, we attach the current working directory, else we use it as is, right?

Comment (incertum) [reply userspace/libsinsp/sinsp_filtercheck_fspath.cpp]: > else we use it as is, right?

almost correct we still need to resolve the absolute path as absolute paths can still contain path traversal. Statement based on real data observed.

**Review by LucaGuerra — APPROVED** [review]
LGTM

**Review by FedeDP — APPROVED** [review]
/approve
